### PR TITLE
chore: added route listener on non-trackable URLs

### LINKED ISSUE #24892: [Bug]: Browser back button does not work
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

While pressing the browser back, the url is changing but the app does not respond to the changes

### Steps To Reproduce

1. Click on multiple entities on the entity explorer
2. Press the back button
3. Notice the app does not go back to the older entity

### Environment

Production

### Issue video log

https://github.com/appsmithorg/appsmith/assets/12022471/79219b8e-1116-405e-a1e3-a827c903e6a5

### Version

v1.9.15

### PR DESCRIPTION
**Pull request by ankitsrivas14**
Added route listener using NavigationSaga for non-trackable urls to track and send usage pulse.
>
>
#### PR fixes following issue(s)
Fixes #24892 
>
>
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
- [x] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/ce/sagas/NavigationSagas.ts [hunk 1]
```diff
@@ -65,6 +67,16 @@ function* clearErrors() {
   }
 }
 
+function* watchForTrackableUrl() {
+  const isTrackable: boolean = yield call(
+    UsagePulse.isTrackableUrl,
+    window.location.pathname,
+  );
+  if (!isTrackable) {
```
Comment (sbalaji1192) [inline app/client/src/ce/sagas/NavigationSagas.ts:75]: We should call track when isTrackable is true right?

Comment (sbalaji1192) [reply app/client/src/ce/sagas/NavigationSagas.ts:75]: And also no need to check UsagePulse.isTrackableUrl here. track function checks that before sending pulse

**Review by sbalaji1192 — COMMENTED** [review]
My suggestion would be to keep the usage pulse as independent of the application as possible for better maintenance. But if we cannot figure out the problem with history listen or window pop state event is not working for then we can go this route.

File: app/client/src/ce/sagas/NavigationSagas.ts (hunk 1 above)
Comment (ankitsrivas14) [reply app/client/src/ce/sagas/NavigationSagas.ts:75]: @sbalaji1192 - If we do not use `UsagePulse.isTrackableUrl` here then moving from Trackable URL to Trackable URL will also send a pulse, which means each time user clicks on a widget, a pulse would fire.

**Comment by ankitsrivas14** [comment]
@sbalaji1192 - QA verified on https://github.com/appsmithorg/appsmith-ee/pull/1966#issuecomment-1664961205
Please check.

**Comment by ankitsrivas14** [comment]
@RakshaKShetty - Back Button and Usage Pulse QA for CE was also done, right? Please confirm.

**Comment by RakshaKShetty** [comment]
@ankitsrivas14 yes its done from my side

**Review by dipyamanbiswas07 — APPROVED** [review]

**Comment by ankitsrivas14** [comment]
After discussing with @sbalaji1192 and @dipyamanbiswas07 regarding this ticket. We have decided to merge this PR as it will fix https://github.com/appsmithorg/appsmith/issues/24892 and the Usage pulse would remain as it is (QA verified).
We have created https://github.com/appsmithorg/appsmith/issues/25617 ticket to look into `history.listen` as per the bandwidth.

**Comment by sbalaji1192** [comment]
@ankitsrivas14 Can you please remove the dead code before merging the pr?

**Comment by ankitsrivas14** [comment]
I've removed the dead code and updated the logic to send pulse only when
1. App mounted on a trackable URL
2. Route changes from non-trackable URL to trackable URL

Please check @dipyamanbiswas07 and @sbalaji1192

**Review by dipyamanbiswas07 — APPROVED** [review]

**Comment by RakshaKShetty** [comment]
@ankitsrivas14 this looks good

**Pull request merged**